source/code/plugin/health/health_monitor_provider.rb

```diff
@@ -0,0 +1,123 @@
+                monitor_labels.keys.each do |key|
+                    labels[key] = monitor_labels[key]
+                end
+            end
+
+            prev_records = health_monitor_instance_state.prev_records
+            time_first_observed = health_monitor_instance_state.state_change_time # the oldest collection time
+            new_state = health_monitor_instance_state.new_state # this is updated before formatRecord is called
+            old_state = health_monitor_instance_state.old_state
+
+            config = get_config(monitor_id)
+
+            if prev_records.size == 1
+                details = prev_records[0]
+            else
+                details = prev_records
+            end
+
+            time_observed = Time.now.utc.iso8601
+
+            monitor_record = {}
+
+            monitor_record[HealthMonitorRecordFields::CLUSTER_ID] = @cluster_id
+            monitor_record[HealthMonitorRecordFields::MONITOR_LABELS] = labels.to_json
+            monitor_record[HealthMonitorRecordFields::MONITOR_ID] = monitor_id
+            monitor_record[HealthMonitorRecordFields::MONITOR_INSTANCE_ID] = monitor_instance_id
+            monitor_record[HealthMonitorRecordFields::NEW_STATE] = new_state
+            monitor_record[HealthMonitorRecordFields::OLD_STATE] = old_state
+            monitor_record[HealthMonitorRecordFields::DETAILS] = details.to_json
+            monitor_record[HealthMonitorRecordFields::MONITOR_CONFIG] = config.to_json
+            monitor_record[HealthMonitorRecordFields::AGENT_COLLECTION_TIME] = Time.now.utc.iso8601
+            monitor_record[HealthMonitorRecordFields::TIME_FIRST_OBSERVED] = time_first_observed
+
+            return monitor_record
+        end
+
+        def get_config(monitor_id)
+            if @monitor_configuration.key?(monitor_id)
+                return @monitor_configuration[monitor_id]
+            else
+                return {}
+            end
+        end
+
+        def get_labels(health_monitor_record)
+            monitor_labels = Hash.new
+            @cluster_labels.keys.each{|key|
+                monitor_labels[key] = @cluster_labels[key]
+            }
+            monitor_id = health_monitor_record[HealthMonitorRecordFields::MONITOR_ID]
+            case monitor_id
+            when HealthMonitorConstants::CONTAINER_CPU_MONITOR_ID, HealthMonitorConstants::CONTAINER_MEMORY_MONITOR_ID, HealthMonitorConstants::USER_WORKLOAD_PODS_READY_MONITOR_ID, HealthMonitorConstants::SYSTEM_WORKLOAD_PODS_READY_MONITOR_ID
+
+                namespace = health_monitor_record[HealthMonitorRecordFields::DETAILS]['details']['namespace']
+                workload_name = health_monitor_record[HealthMonitorRecordFields::DETAILS]['details']['workloadName']
+                workload_kind = health_monitor_record[HealthMonitorRecordFields::DETAILS]['details']['workloadKind']
+
+                monitor_labels['container.azm.ms/workload-name'] = workload_name.split('~~')[1]
```

Comment:
check allowed character set for k8s object names

Reply:
Yes. only dots and dashes